contributing/development/core_and_modules/internal_audio_architecture.rst

```diff
@@ -0,0 +1,236 @@
+    but is smaller in file size than WAV and MP3. Therefore most suitable for playing music in-game.
+
+
+**Why not vorbisfile?**
+// TODO
+
+
+WAV
+^^^
+
+.. note::
+
+    This is an audio format that is the qickest to load, but is larger in file size than Ogg Vorbis and MP3. Therefore most suitable for playing sound effects.
+
+
+MP3
+^^^
+
+
+.. _doc_internal_audio_architecture_servers:
+
+AudioServer
+------------
+
+Decoding
+^^^^^^^^
+
+Buffers
+^^^^^^^
+
+Kinds of buffers
+- main buffer: is hardcoded to `512`
+
+   - But the driver buffer size calculation is unique to each driver and depends on varying factors, such as:
+
+      - but depends in general on: audio output latency, mix rate, channel count (which depends on the speaker count)
+
+- lookahead buffer: We store the next few samples so we have some time to fade audio out if it ends abruptly at the beginning of the next mix.
+- mix buffer: sum of main buffer and the lookahead buffer
+
+
+The actual audio frames are written into the channels, which are split into left and right stereo pairs (i.e. physical speakers).
+
+
+Mixing
+^^^^^^
+
+**Playback**
+- Pausing triggers a fade out.
+- Stopping a playing stream triggers a fade out as well and frees the affiliated memory.
+
+
+**Fading**
+- A fade is achieved by lerping the audio volume over multiple frames.
+
+
+**Channels**
+The channel count depends on the amount of speakers:
+- stereo: 1
+
+   - front-left speaker
+   - front-right speaker
+
+- surround 3.1: 2
+
+   - above speakers, plus:
+   - center speaker
+   - LFE(?) (always full volume)
+
+- surround 5.1: 3
+
+   - above speakers, plus:
+   - rear-left speaker
+   - rear-right speaker
+
+- surround 7.1: 4
+
+   - above speakers, plus:
+   - side-left speaker
+   - side-right speaker
+
+But those are not the total channels, as each has a left and right side (i.e. left and right speaker).
+The total channels are twice the size mentioned above and equals the number of speakers.
+Depending on the context, one or the other size logic is used.
+
+The speaker mode is sometimes driver-dependent. All drivers lookup the mode by using the total number of channels.
+Except Alsa, XAudio2 and OpenSL ES. These enforce stereo speaker mode and support only 2 total channels.
+
+
+**Bus**
+Godot as an internal bus logic to separately control volume and audio effects applied to audio streams.
+All busses eventually feed into the master bus, which is output to the audio output device.
+
+
+Seeking / Looping
+^^^^^^^^^^^^^^^^^
+
+
+SpatialSoundServer
+^^^^^^^^^^^^^^^^^^
+
+
+SpatialSound2DServer
+^^^^^^^^^^^^^^^^^^^^
+
+
+.. _doc_internal_audio_architecture_classes:
+
+Core audio classes architecture
+-------------------------------
+
+
+.. _doc_internal_audio_architecture_2d_vs_3d:
+
+
+2D and 3D audio separation
+--------------------------
+
+
+.. _doc_internal_audio_architecture_techniques:
+
+
+2D audio techniques
+-------------------
+
+
+3D audio techniques
+-------------------
+
```

Comment:
It's a good idea to mention SPCAP, Camera3D doppler tracking and the audio panning properties here.

Reply:
Copy&pasted that as-is into the file for now. I'll write that part today.